Preserve syntax highlighting while editing

ArsenijeD opened the pull request:
Handle new line in a way that it behaves exactly the same as 'space' character.

ArsenijeD committed:
Preserve syntax highlighting while editing

Handle new line in a way that it behaves exactly
the same as 'space' character.

ArsenijeD commented:
@vermorel 

This should fix issue related to coloring mess-up every time `.nvn` document is edited.

I think that the best solution we have here is to modify `tokenize(line: string, state: IState): ILineTokens` function inside `envisionHighlighter.ts` file by handling additional `TokenType`, which is basically `new line`.

On the other side, since you were strict about not modifying file mentioned (it's part of another Lokad project), I had to fix `_parseTokens` method inside `EnvisionSemanticTokensProvider` by handling new line in a same way as it's just one space character.

Please let me know your opinion on this!

vermorel commented:
Not touching the `tokenize` function was the right move. Thanks!

I have also validated that the fix works. I would only request to add some minimal inline doc to clarify why this code is introduced (see code comment), and then we are good.

vermorel commented on src/envisionSemanticTokensProvider.ts:
```diff
@@ -63,33 +63,46 @@ class EnvisionSemanticTokensProvider implements vscode.DocumentSemanticTokensPro
-
-                tokens.push(0);  // No modifiers for now, hence 0
-                prevStart = token.startIndex;
+            if (isLineEmpty) {
```
@ArsenijeD This works, but it would be nicer to add an inline comment to outline why this logic is needed in the first place. In particular, do you understand why VS Code appears to be differing in behavior from Monaco?

ArsenijeD replied:
I've just added comment!

To answer your question: not very likely. There shouldn't be any difference, since VSCode relies on Monaco editor. Anyways, additional investigation on that topic could give us an answer.

ArsenijeD committed:
Add inline explanation

Explain while additional token is introduced, when it comes to facing empty line in the .nvn file.
